{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9262235f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from scipy.stats import stats\n",
    "from tqdm.notebook import trange\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.ticker import MultipleLocator, AutoLocator\n",
    "\n",
    "from function import conserve\n",
    "from function.utilities import get_taxid_dict\n",
    "from function.utilities import get_protein_name\n",
    "from function.utilities import get_fasta_seq_info\n",
    "from function.utilities import get_uniprot_rawdata\n",
    "from function.utilities import find_human_sequence\n",
    "from function.seqfilter import SeqFilter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cf5a4c8",
   "metadata": {},
   "source": [
    "# Param"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "54ba3650",
   "metadata": {},
   "outputs": [],
   "source": [
    "#####CHANGE HERE#####\n",
    "algorithm = \"VSL2\" #VLXT, VSL2, VL3-BA, fidpnn\n",
    "uniprot_id = \"Q13148\" #'Q13148','P35637', 'P49790', 'Q8NDV7'\n",
    "conserve_method = 'jsd' #['shannon_entropy','jsd','r4s']\n",
    "#####CHANGE HERE#####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eb39f28b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#see tax\n",
    "tax_ids = [7711, 7742, 32523, 40674] #chordata, vertebrata, tetrapoda, mammalia\n",
    "\n",
    "#order/disorder length criteria\n",
    "order_filter_length = 10\n",
    "disorder_filter_length = 40\n",
    "\n",
    "#conserve_method config, only for shannon_entropy and jsd\n",
    "#same as https://compbio.cs.princeton.edu/conservation/score.html\n",
    "normalize = True \n",
    "windows_size = 0\n",
    "sequence_weight = False\n",
    "\n",
    "#fig2b amino acid show order\n",
    "conserve_peraa_sort = [\"A\",\"C\",\"D\",\"E\",\"G\",\"H\",\"I\",\"K\",\"L\",\"M\",\"N\",\"P\",\"Q\",\"R\",\"S\",\"T\",\"V\",\"F\",\"W\",\"Y\"]\n",
    "\n",
    "#path\n",
    "input_path = Path(\"./output/fasta/d_extre_filtered\")\n",
    "fig2_path = Path(\"./output/fig2\") / Path(conserve_method) #for figure 2\n",
    "table_homology_path = Path(\"./output/table_homology\") #for homologous table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6970228",
   "metadata": {},
   "source": [
    "# Load uniprot and order/disorder identification data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d3396d89",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_uniprot_path = Path(\"./rawdata/human_uniprot.tab\")\n",
    "human_df = get_uniprot_rawdata(human_uniprot_path)\n",
    "\n",
    "pondr_disorder_ident_path = Path(\"./output/pondr_disorder_ident\")\n",
    "od_human_df = pd.read_pickle(pondr_disorder_ident_path / \"{}_od_human_df.pkl\".format(algorithm))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "680fd2d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fff4056",
   "metadata": {},
   "source": [
    "# Conservation Score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49ba9afa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_only_human_score(score, fasta_path):\n",
    "    \"\"\"\n",
    "    get entropy score when alignment positions on human sequences aren't gap\n",
    "\n",
    "    score: list, entropy score calculated by get_conserve_score()\n",
    "    fasta_path: fasta file path\n",
    "    \n",
    "    return: list, score as same length with human sequence\n",
    "    \"\"\"\n",
    "    nogap_score = []\n",
    "    nogap = np.array(list(find_human_sequence(fasta_path)[\"sequence\"])) != \"-\"\n",
    "    nogap_index = np.where(nogap)[0].tolist()\n",
    "    for index in nogap_index:\n",
    "        nogap_score.append(score[index])\n",
    "        \n",
    "    return nogap_score\n",
    "\n",
    "def get_normalize_by_chordata(conserve_list):\n",
    "    \"\"\"\n",
    "    get normalized score while mean and std is derived from chordata sequence\n",
    "    \n",
    "    conserve_list: list, list of conservation info from get_conserve_score()\n",
    "    \n",
    "    return: list, list of conservation info with normalized score\n",
    "    \"\"\"\n",
    "    #get mean and std\n",
    "    z_score_base = np.array(conserve_list[0]['score'])\n",
    "    mean = z_score_base.mean()\n",
    "    std = z_score_base.std()\n",
    "    \n",
    "    #get z-score\n",
    "    score = np.stack([i['score'] for i in conserve_list])\n",
    "    score = (score - mean) / std\n",
    "    \n",
    "    #put back to conserve_list\n",
    "    for index, element in enumerate(score):\n",
    "        element = element.tolist()\n",
    "        conserve_list[index]['score'] = element\n",
    "        \n",
    "    return conserve_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bec9d469",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_conserve_score(tax_id, uniprot_id):\n",
    "    \"\"\"\n",
    "    get conserve score with many info for plot\n",
    "    \n",
    "    tax_id: int, taxonomy id, i.e. 7711\n",
    "    uniprot_id: str, uniprot id\n",
    "    \n",
    "    return: dict, conserve score with many info\n",
    "    \"\"\"\n",
    "    #read fasta\n",
    "    fasta_path = input_path / str(tax_id) / \"{}.fasta\".format(uniprot_id)\n",
    "\n",
    "    #get conserve score\n",
    "    conserve_score = conservefunc.get_conserve_score(fasta_path, \n",
    "                                                     method=conserve_method, \n",
    "                                                     windows_size=windows_size,\n",
    "                                                     sequence_weight=sequence_weight)\n",
    "\n",
    "    #only nogap conserve_score, r4s removes gap score base on reference sequence(human sequence)\n",
    "    if conserve_method != 'r4s': \n",
    "        conserve_score = get_only_human_score(conserve_score, fasta_path)\n",
    "\n",
    "    #get order disorder region by PONDR algorithm\n",
    "    od_ident = od_human_df[od_human_df[\"uniprot_id\"] == uniprot_id][\"od_ident\"].values[0]  \n",
    "    od_ident = seqfilter.length_filter_by_od_ident(od_ident=od_ident, \n",
    "                                                   disorder_filter_length=disorder_filter_length, \n",
    "                                                   order_filter_length=order_filter_length)\n",
    "\n",
    "    #conservation score and od_ident length error handle, \n",
    "    #due to different sequence record between OMA and uniprot but their uniprot_ids are the same\n",
    "    if len(conserve_score) != len(od_ident):\n",
    "        raise Exception(\"{} lenth of conservation score is not equal with od_ident\".format(uniprot_id))\n",
    "        \n",
    "    #get disorder region after length check\n",
    "    region = seqfilter.get_od_index(od_ident)\n",
    "\n",
    "    #get human sequence without gap for disorder plot\n",
    "    human_sequence = find_human_sequence(fasta_path)[\"remove_gap_sequence\"]\n",
    "\n",
    "    #get infos\n",
    "    taxonomy = get_taxid_dict()[tax_id]\n",
    "    gene_name = get_protein_name(uniprot_id, human_df)[\"gene_name\"]\n",
    "    homologous_info = get_fasta_seq_info(fasta_path)[\"homologous_info\"]\n",
    "\n",
    "    return {\n",
    "        \"uniprot_id\": uniprot_id,\n",
    "        \"gene_name\": gene_name,\n",
    "        \"tax_id\": tax_id,\n",
    "        \"taxonomy\": taxonomy,\n",
    "        \"score\": conserve_score,\n",
    "        \"od_ident\": od_ident,\n",
    "        \"human_sequence\": human_sequence,\n",
    "        \"region\": region,\n",
    "        \"homologous_info\":homologous_info\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61a45b58",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_conserve_peraa_score(conserve_list):\n",
    "    \"\"\"\n",
    "    get conserve score per amino acid\n",
    "    \n",
    "    conserve_list: list, conserve info by get_conserve_score()\n",
    "    \n",
    "    return: dict, conserve score with many info\n",
    "    \"\"\"\n",
    "    #array's dim: 2(aa content/peraa conserve), 2(order/disorder), 20(A, C, D, E...), 4(taxonomy id)\n",
    "    peraa_score = np.zeros((2, 2, 20, len(conserve_list)))\n",
    "    \n",
    "    taxonomy_list = []\n",
    "    \n",
    "    for tax_index, i in enumerate(conserve_list):\n",
    "        \n",
    "        #make tax list\n",
    "        taxonomy_list.append(i['taxonomy'])\n",
    "        \n",
    "        #read fasta\n",
    "        fasta_path = input_path / str(i['tax_id']) / \"{}.fasta\".format(i['uniprot_id'])\n",
    "        \n",
    "        #get conserve score and aa content\n",
    "        info_dict = conserveperaa.get_aa_info(fasta_path, i['score'], i['od_ident'])\n",
    "        order_conserve, order_content = (info_dict[\"conserve\"][\"order\"], info_dict[\"content\"][\"order\"])\n",
    "        disorder_conserve, disorder_content = (info_dict[\"conserve\"][\"disorder\"], info_dict[\"content\"][\"disorder\"])\n",
    "        \n",
    "        #put into array\n",
    "        for aa_index, element in enumerate(conserve_peraa_sort):\n",
    "            peraa_score[0][0][aa_index][tax_index] = order_content[element]\n",
    "            peraa_score[0][1][aa_index][tax_index] = disorder_content[element]\n",
    "            peraa_score[1][0][aa_index][tax_index] = order_conserve[element]\n",
    "            peraa_score[1][1][aa_index][tax_index] = disorder_conserve[element]\n",
    " \n",
    "    return {\n",
    "        \"uniprot_id\": i['uniprot_id'],\n",
    "        \"gene_name\": i['gene_name'],\n",
    "        \"taxonomy\": taxonomy_list,\n",
    "        \"conserve_peraa_sort\": conserve_peraa_sort,\n",
    "        \"peraa_score\": peraa_score,\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7675edff",
   "metadata": {},
   "outputs": [],
   "source": [
    "#init some func\n",
    "seqfilter = SeqFilter()\n",
    "conserveperaa = conserve.ConservePeraa()\n",
    "if conserve_method == 'jsd' or conserve_method == 'shannon_entropy':\n",
    "    #source code: Capra JA and Singh M. Predicting functionally important residues from sequence conservation.\n",
    "    #             Bioinformatics, 23(15):1875-82, 2007. [Bioinformatics] \n",
    "    #by web crawler\n",
    "    \n",
    "    #conservefunc = conserve.ConserveByWeb()\n",
    "    \n",
    "    #standalone version\n",
    "    conservefunc = conserve.ConserveStandalone()\n",
    "     \n",
    "elif conserve_method == 'r4s':\n",
    "    conservefunc = conserve.Rate4Site()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0923fadb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calcuate conserve score given tax_id\n",
    "conserve_list = []\n",
    "\n",
    "t = trange(len(tax_ids), leave=True)\n",
    "for i in t:\n",
    "    t.set_description(str(tax_ids[i]))\n",
    "    t.refresh()\n",
    "\n",
    "    conserve_score = get_conserve_score(tax_id=tax_ids[i], uniprot_id=uniprot_id)\n",
    "    conserve_list.append(conserve_score)\n",
    "    \n",
    "#conserce score normalization\n",
    "if normalize:\n",
    "    conserve_list = get_normalize_by_chordata(conserve_list)\n",
    "    \n",
    "#calculate peraa score \n",
    "conserve_peraa = get_conserve_peraa_score(conserve_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30da9a3f",
   "metadata": {},
   "source": [
    "# Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41f66bce",
   "metadata": {},
   "outputs": [],
   "source": [
    "f, (a0) = plt.subplots(1, 1, sharex=True, sharey=True, figsize=(20, 3.63), dpi=200)\n",
    "\n",
    "for i in conserve_list:\n",
    "\n",
    "    # line\n",
    "    if i[\"taxonomy\"] == \"chordata\":\n",
    "        a0.plot(i[\"score\"], color=\"#335079\", zorder=100, linewidth=3)\n",
    "    elif i[\"taxonomy\"] == \"vertebrata\":\n",
    "        a0.plot(i[\"score\"], color=\"#b5b5b5\", zorder=2, linewidth=0.9)\n",
    "    elif i[\"taxonomy\"] == \"tetrapoda\":\n",
    "        a0.plot(i[\"score\"], color=\"#b5b5b5\", zorder=2, linewidth=0.9)\n",
    "    elif i[\"taxonomy\"] == \"mammalia\":\n",
    "        a0.plot(i[\"score\"], color=\"#63a2fe\", zorder=90, linewidth=3)\n",
    "\n",
    "#disorder plot\n",
    "for region in conserve_list[0][\"region\"][\"disorder_region\"]:\n",
    "    start = region[\"start\"]\n",
    "    start = start / len(conserve_list[0][\"human_sequence\"])\n",
    "    end = region[\"end\"]\n",
    "    end = end / len(conserve_list[0][\"human_sequence\"])\n",
    "    a0.axhline(-3, start, end, color=\"#ff0000\", linewidth=3, zorder=200)\n",
    "    \n",
    "#0 line\n",
    "a0.axhline(0 , color=\"#a2a2a2\", linewidth=0.5, zorder=100)\n",
    "a0.axhline(1 , color=\"#a2a2a2\", linewidth=0.5, linestyle='--', zorder=100)\n",
    "a0.axhline(-1, color=\"#a2a2a2\", linewidth=0.5, linestyle='--', zorder=100)\n",
    "\n",
    "#WYF point\n",
    "for index, (od, aa, score) in enumerate(\n",
    "    zip(conserve_list[0][\"od_ident\"], conserve_list[0][\"human_sequence\"], conserve_list[0][\"score\"])):\n",
    "    if od == \"1\":\n",
    "        if aa == \"Y\":\n",
    "            a0.scatter(index, score, s=320, color=\"#ffff00\", edgecolors=\"black\", linewidths=0.5, zorder=200)\n",
    "        elif aa == \"F\":\n",
    "            a0.scatter(index, score, s=320, color=\"#ffa300\", edgecolors=\"black\", linewidths=0.5, zorder=200)\n",
    "        elif aa == \"W\":\n",
    "            a0.scatter(index, score, s=320, color=\"#9400d1\", edgecolors=\"black\", linewidths=0.5, zorder=200)\n",
    "\n",
    "#style\n",
    "a0.set_xlim(0, len(i[\"score\"]))\n",
    "a0.set_ylim(-3.9, 3)\n",
    "\n",
    "a0.xaxis.set_major_locator(MultipleLocator(100))\n",
    "a0.xaxis.set_minor_locator(MultipleLocator(20))\n",
    "a0.yaxis.set_major_locator(MultipleLocator(1))\n",
    "# a0.yaxis.set_minor_locator(MultipleLocator(0.5))\n",
    "\n",
    "a0.tick_params(which='major', axis=\"x\",direction=\"in\", length=5)\n",
    "a0.tick_params(which='minor', axis=\"x\",direction=\"in\", length=2)\n",
    "a0.tick_params(which='major', axis=\"y\",direction=\"in\", length=5)\n",
    "a0.tick_params(which='minor', axis=\"y\",direction=\"in\", length=2)\n",
    "\n",
    "a0.axes.xaxis.set_ticklabels([])\n",
    "a0.axes.yaxis.set_ticklabels([])\n",
    "\n",
    "title = \"{} ({})\".format(conserve_list[0][\"gene_name\"], conserve_list[0][\"uniprot_id\"], fontsize=12)\n",
    "# a0.set_title(title, loc=\"left\")\n",
    "# a0.set_xlabel(\"Residue Number\", fontsize=12)\n",
    "# a0.set_ylabel(\"Conservation Level (Z-score)\", fontsize=12)\n",
    "\n",
    "plt.tight_layout(pad=0.3)\n",
    "\n",
    "#savefig\n",
    "# path = fig2_path / \"{}_{}_fig2up\".format(conserve_list[0][\"gene_name\"], conserve_list[0][\"uniprot_id\"])\n",
    "# f.savefig(\"{}.svg\".format(str(path)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "054107d0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#array's dim: 2(aa content/peraa conserve), 2(order/disorder), 20(A, C, D, E...), 4(taxonomy id)\n",
    "f, (order,disorder) = plt.subplots(2, 20, sharex=True, sharey=True, figsize=(20, 2.2), dpi=200)\n",
    "\n",
    "y_max = np.nanmax(conserve_peraa['peraa_score'][1]) + 0.3\n",
    "y_min = np.nanmin(conserve_peraa['peraa_score'][1]) - 0.5\n",
    "\n",
    "for index, sub in enumerate(order):\n",
    "    sub.plot(conserve_peraa['peraa_score'][1][0][index], linestyle=\"solid\", color=\"#a2a2a2\",linewidth=3)\n",
    "    sub.set_xlim(-0.5, 3.5)\n",
    "    sub.set_ylim(y_min, y_max)\n",
    "\n",
    "#     sub.axes.yaxis.set_ticklabels([])\n",
    "    sub.axes.xaxis.set_ticklabels([])\n",
    "    sub.get_xaxis().set_visible(False)\n",
    "    \n",
    "    sub.tick_params(which='major', axis=\"y\",direction=\"in\", length=2)\n",
    "    sub.tick_params(which='minor', axis=\"y\",direction=\"in\", length=1)\n",
    "    \n",
    "    sub.yaxis.set_major_locator(MultipleLocator(1))\n",
    "    sub.yaxis.set_minor_locator(MultipleLocator(0.5))\n",
    "    \n",
    "    yticks = sub.yaxis.get_major_ticks()\n",
    "#     yticks[1].label.set_visible(False)\n",
    "    \n",
    "for index, sub in enumerate(disorder):\n",
    "    sub.plot(conserve_peraa['peraa_score'][1][1][index], linestyle=\"solid\", color=\"#a2a2a2\",linewidth=3)\n",
    "    \n",
    "#     sub.axes.yaxis.set_ticklabels([])\n",
    "    sub.axes.xaxis.set_ticklabels([])\n",
    "    sub.get_xaxis().set_visible(False)\n",
    "    \n",
    "    sub.tick_params(which='major', axis=\"y\",direction=\"in\", length=2)\n",
    "    sub.tick_params(which='minor', axis=\"y\",direction=\"in\", length=1)\n",
    "    \n",
    "    sub.yaxis.set_major_locator(MultipleLocator(1))\n",
    "    sub.yaxis.set_minor_locator(MultipleLocator(0.5))\n",
    "    \n",
    "    yticks = sub.yaxis.get_major_ticks()\n",
    "#     yticks[1].label.set_visible(False)\n",
    "    \n",
    "plt.subplots_adjust(wspace=0.1)\n",
    "plt.tight_layout(pad=0.3)\n",
    "\n",
    "#savefig\n",
    "# path = fig2_path / \"{}_{}_fig2down\".format(conserve_list[0][\"gene_name\"], conserve_list[0][\"uniprot_id\"])\n",
    "# f.savefig(\"{}.svg\".format(str(path)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19e3489f",
   "metadata": {},
   "source": [
    "# Output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fc85a93",
   "metadata": {},
   "outputs": [],
   "source": [
    "#homologous output\n",
    "path = table_homology_path / \"{}_{}_homologous.xlsx\".format(conserve_list[0][\"gene_name\"], conserve_list[0][\"uniprot_id\"])\n",
    "with pd.ExcelWriter(path) as writer:\n",
    "    for i in conserve_list:\n",
    "        df = pd.DataFrame(columns=['oma_protein_id','species','taxon_id','oma_cross_reference'])\n",
    "        taxonomy = i['taxonomy']\n",
    "        for j in i['homologous_info']:\n",
    "            df = df.append({'oma_protein_id': j['oma_protein_id'],\n",
    "                            'species': j['species'],\n",
    "                            'taxon_id': j['taxon_id'],\n",
    "                            'oma_cross_reference': j['oma_cross_reference']},ignore_index=True)\n",
    "        \n",
    "        df.to_excel(writer, sheet_name=taxonomy)  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "759eada9",
   "metadata": {},
   "source": [
    "# Overall trend"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73879c89",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# from function.utilities import get_subset\n",
    "# input_path = Path(\"/home/wenlin/d/rbp/oma_all/e_extre_filtered_rbp\")\n",
    "# uniprot_ids = get_subset(human_df, 'rbp')['uniprot_id'].tolist() \n",
    "\n",
    "# all_list = []\n",
    "# failed_uids = []\n",
    "# j = trange(len(uniprot_ids), leave=True)\n",
    "# for uniprot_id in j:\n",
    "#     try:\n",
    "#         conserve_list = []\n",
    "\n",
    "#         t = trange(len(tax_ids), leave=False)\n",
    "#         for i in t:\n",
    "#             t.set_description(str(tax_ids[i]))\n",
    "#             t.refresh()\n",
    "\n",
    "#             conserve = get_conserve_score(tax_id=tax_ids[i], uniprot_id=uniprot_ids[uniprot_id])\n",
    "#             conserve_list.append(conserve)\n",
    "\n",
    "\n",
    "#         #conserce score normalize\n",
    "#         if normalize:\n",
    "#             conserve_list = get_normalize_by_chordata(conserve_list)\n",
    "\n",
    "#         #calculate peraa score \n",
    "#         conserve_peraa = get_conserve_peraa_score(conserve_list)\n",
    "\n",
    "\n",
    "#         all_list.append({\"uniprot_id\":conserve_peraa['uniprot_id'],\n",
    "#                          \"gene_name\":conserve_peraa['gene_name'],\n",
    "#                          \"peraa_score\":conserve_peraa['peraa_score']})\n",
    "        \n",
    "#     except Exception as e: print(e)\n",
    "\n",
    "# no_zero_all_list = []\n",
    "# for i in all_list:\n",
    "#     if not (i['peraa_score'][0][0] == 0).all():\n",
    "#         no_zero_all_list.append(i)       \n",
    "# order_conserve = np.stack([i['peraa_score'][1][0] for i in no_zero_all_list])\n",
    "# order_conserve = np.nanmean(order_conserve,axis=0)\n",
    "# # order_conserve = order_conserve.mean(axis=0)\n",
    "\n",
    "\n",
    "# no_zero_all_list = []\n",
    "# for i in all_list:\n",
    "#     if not (i['peraa_score'][0][1] == 0).all():\n",
    "#         no_zero_all_list.append(i)       \n",
    "# disorder_conserve = np.stack([i['peraa_score'][1][1] for i in no_zero_all_list])\n",
    "# disorder_conserve = np.nanmean(disorder_conserve,axis=0)\n",
    "# # disorder_conserve = disorder_conserve.mean(axis=0)\n",
    "\n",
    "# overall_conserve = np.stack([order_conserve, disorder_conserve])\n",
    "# np.save('overall_conserve.pkl',overall_conserve)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8c49b5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "overall_trend_path = Path(\"./output/peraa_overall_trend\")\n",
    "overall_conserve = np.load('./output/peraa_overall_trend/overall_conserve.pkl.npy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5352343",
   "metadata": {},
   "outputs": [],
   "source": [
    "f, (order,disorder) = plt.subplots(2, 20, sharex=True, sharey=True, figsize=(20, 2.2), dpi=200)\n",
    "\n",
    "y_max = np.nanmax(overall_conserve) + 0.15\n",
    "y_min = np.nanmin(overall_conserve) - 0.15\n",
    "\n",
    "for index, sub in enumerate(order):\n",
    "    sub.plot(overall_conserve[0][index], linestyle=\"solid\", color=\"#a2a2a2\",linewidth=3)\n",
    "    \n",
    "\n",
    "#     sub.axes.yaxis.set_ticklabels([])\n",
    "    sub.axes.xaxis.set_ticklabels([])\n",
    "    sub.get_xaxis().set_visible(False)\n",
    "    \n",
    "    sub.tick_params(which='major', axis=\"y\",direction=\"in\", length=2)\n",
    "    sub.tick_params(which='minor', axis=\"y\",direction=\"in\", length=1)\n",
    "    \n",
    "    sub.yaxis.set_major_locator(MultipleLocator(1))\n",
    "    sub.yaxis.set_minor_locator(MultipleLocator(0.5))\n",
    "    \n",
    "    yticks = sub.yaxis.get_major_ticks()\n",
    "#     yticks[1].label.set_visible(False)\n",
    "    \n",
    "    \n",
    "for index, sub in enumerate(disorder):\n",
    "    sub.plot(overall_conserve[1][index], linestyle=\"solid\", color=\"#a2a2a2\",linewidth=3)\n",
    "    sub.set_xlim(-0.5, 3.5)\n",
    "    sub.set_ylim(y_min, y_max)\n",
    "    \n",
    "    \n",
    "#     sub.axes.yaxis.set_ticklabels([])\n",
    "    sub.axes.xaxis.set_ticklabels([])\n",
    "    sub.get_xaxis().set_visible(False)\n",
    "    \n",
    "    sub.tick_params(which='major', axis=\"y\",direction=\"in\", length=2)\n",
    "    sub.tick_params(which='minor', axis=\"y\",direction=\"in\", length=1)\n",
    "    \n",
    "    sub.yaxis.set_major_locator(MultipleLocator(1))\n",
    "    sub.yaxis.set_minor_locator(MultipleLocator(0.5))\n",
    "    \n",
    "    yticks = sub.yaxis.get_major_ticks()\n",
    "#     yticks[1].label.set_visible(False)\n",
    "    \n",
    "plt.subplots_adjust(wspace=0.1)\n",
    "plt.tight_layout(pad=0.3)\n",
    "\n",
    "\n",
    "path = overall_trend_path / \"overall_trend\"\n",
    "f.savefig(\"{}.svg\".format(str(path)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aed96aae",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
